src/index.d.ts

```diff
@@ -0,0 +1,120 @@
+    // An action can be given a payload.
+    | [Action<S, P, D>, Payload<P>]
+    // State can be set directly along with any effects to run.
+    // This is particularly when setting the initial state.
+    | [State<S>, ...EffectDescriptor<D>[]]
+    | State<S>
+
+  // A payload allows data not within the state to be given to a dispatch.
+  type Payload<P> = P
+
+  // An effect descriptor describes how an effect should be invoked.
+  // A function that creates this is called an effect constructor.
+  type EffectDescriptor<D> = [Effect, EffectData<D>]
+
+  // An effect is where side effects and any additional dispatching occur.
+  type Effect = <D>(dispatch: Dispatch, props?: EffectData<D>) => void
+
+  // An effect is generally given additional data.
+  type EffectData<D> = D
+
+  // -----------------------------------------------------------------------------
+
+  // A virtual DOM node represents an actual DOM element.
+  type VDOM = {
+    readonly type: string;
+    readonly props: PropList;
+    readonly children: VNode[];
+    node: MaybeNode;
+    readonly tag?: Tag;
+    readonly key: Key;
+    memo?: PropList;
+  }
+
+  // Virtual DOM properties will often correspond to HTML attributes.
+  type PropList = Record<string, Prop> & Readonly<{ key?: Key }> & ElementCreationOptions
+  type Prop = bigint | boolean | null | number | string | symbol | undefined | Function | ClassProp | StyleProp
+
+  // The `class` property represents an HTML class attribute string.
+  type ClassProp = string | Record<string, boolean> | ClassProp[]
+
+  // The `style` property represents inline CSS.
+  type StyleProp = Record<string, null | number | string>
+
+  // A virtual node is a convenience layer over a virtual DOM node.
+  type VNode = null | undefined | VDOM
+
+  // Actual DOM nodes will be manipulated depending on how property patching goes.
+  type MaybeNode = null | undefined | Node
+
+  // A virtual DOM node's tag has metadata relevant to it.
+  type Tag
+    // Virtual DOM nodes are tagged by their type to assist rendering.
+    = VDOMNodeType
+    // Only memoized virtual DOM nodes are tagged with a view.
+    | View
+
+  // Types 1 and 3 are based on actual DOM node types but 2 is Hyperapp-specific.
+  // https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType
+  const enum VDOMNodeType {
+    Recycled = 1,
+    Lazy = 2,
+    Text = 3,
+  }
+
+  // A key can uniquely associate a virtual DOM node with certain a DOM element.
+  type Key = null | string | undefined
+
+  // -----------------------------------------------------------------------------
+
+  // The `app` function initiates a Hyperapp application. `app` along with effects
+  // should be the only places you need to worry about side effects.
+  function app<S, P, D>(props: App<S, P, D>): Dispatch
+
+  // The `h` function builds a virtual DOM node.
+  function h(type: string, props: PropList, children?: VNode | readonly VNode[]): VDOM
```

Comment:
Alternatively i would allow to type the `h` with a generic

```
function h<T = {}>(type: string, props: T & PropList, children?: VNode | readonly VNode[]): VDOM
```

Reply:
Making `h` generic in that way would defeat the purpose of `PropList` if you could just do something like `h<any>`. Please see what I wrote about `Prop` above.